### PR TITLE
docs: rewrite finetuning page as a working, measured recipe

### LINKED ISSUE #3779: OpenAI-compatible providers: fold non-leading system messages for chat templates that reject them (Qwen3.5 on vLLM returns 400 on every gptme request)
## Problem

gptme sends tool results and injected context as `system` messages in the middle of the conversation. Qwen3.5's stock chat template raises `System message must be at the beginning.` for any non-leading system message, so **every** request from gptme to a Qwen3.5 model served by vLLM (or any OpenAI-compatible server that applies that template) fails with HTTP 400. Observed 2026-09-09: 580/580 requests rejected during a gptme-eval run against `Qwen/Qwen3.5-0.8B` on vLLM 0.28.

Other model families with the same template rule will hit this too.

## Existing precedent

`gptme/llm/llm_openai.py` already has `_prep_o1`, which converts system messages (without `call_id`) to `user` messages wrapped in `<system>…</system>` for OpenAI reasoning models that don't support the system role.

## Proposal

- Keep the *first* system message as `system`.
- For OpenAI-compatible/local providers, fold every *non-leading* `system` message (no `call_id`) into a `user` message wrapped in `<system>…</system>`, the same way `_prep_o1` does, controlled by a model-meta flag (e.g. `supports_mid_system=False`) with an env override for custom endpoints (`GPTME_FOLD_SYSTEM_MESSAGES=1`).
- Tool results that are already `tool`-role (native format) are unaffected.

## Workaround used meanwhile

Serve vLLM with a copy of the Qwen3.5 template where the raise is replaced by rendering the message as a ChatML `system` turn. That matches how such data looks in gptme's own logs and in fine-tuning datasets derived from them (see the finetuning docs PR #3778).

Found while running the tool-format fine-tuning experiment; details in the gptme docs PR #3778.

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

Rewrites `docs/finetuning.md` around a recipe that was actually run end to end on
2026-09-08/09, rather than a pointer to a script.

This **supersedes the narrow collector note from #2407**. That PR was the right
call at the time — the old page described a workflow that no longer existed, and
replacing it with an honest description of `scripts/train/collect.py` beat leaving
stale TODOs in published docs. Since then the workflow exists, so the page can be
a recipe again. `collect.py` is still documented, honestly, as the legacy
chat-only collector: it predates gptme's tool formats, so it emits no `tool_calls`
and no `tools` column.

## What the page now covers

- **Data** — `~/.local/share/gptme/logs/*/conversation.jsonl`, the **two** tool-call
  text encodings it carries (markdown fences *and* the native `@tool(call_id): {json}`
  form; matching one silently erases whole sessions), export to Axolotl
  `chat_template` JSONL with JSON-**string** `arguments`, the three renderers and why
  paired datasets are what makes a format comparison mean anything, assistant-only
  masking (~9% trainable tokens on this data shape), time-based holdout, redaction,
  and the `sequence_len` = conversation budget + ~7–8.5k gptme system prompt rule.
- **Training** — why Axolotl over TRL / LLaMA-Factory / torchtune, the Qwen3.5-0.8B
  LoRA config with the three keys that cost money to discover (`micro_batch_size: 1`
  with `sample_packing`, `freeze_mm_modules` for the unified VLM, no LoRA kernels),
  measured numbers (1,170 s on one H100 for 3,080 sessions / ~33M tokens, ~$3 all-in,
  eval loss 1.918 → 1.721), and a model-choice table.
- **Serving + eval** — vLLM flags (`--enable-auto-tool-choice`, `qwen3_xml` vs
  `hermes`, thinking off), then `gptme-eval --model local/<name>@markdown`, plus the
  non-sandboxed-eval prerequisites that otherwise produce a wall of failures that
  reads like a bad model.
- **Results** — base vs markdown-SFT across all three formats, with the caveats
  (n=116, ±4 pp is noise; a 0.8B is weak in absolute terms).
- **Compute** — renting bare GPUs by the hour and the SkyPilot autostop rules that
  stop a forgotten H100 from costing more than the whole experiment.

## The gptme-relevant bug this surfaced

**Qwen3.5's stock chat template raises `System message must be at the beginning`,
and gptme injects context as non-leading `system` messages — so every request to
a vLLM-served Qwen3.5 used to return HTTP 400.** It burned two full eval runs
(580/580 requests failed, both arms scored 0/116) and reads exactly like a
broken model or a timeout; only the server log gives it away.

**Fixed in gptme as of #3780** (merged 2026-09-09, closes #3779). `_fold_mid_system`
keeps the first system message and re-emits later system turns (no `call_id`) as
`<system>` user messages. It activates when the model id contains `qwen3.5` /
`qwen3_5`, or when `GPTME_FOLD_SYSTEM_MESSAGES=1`. This recipe's served names
(`tool-format-experiment`, `sft`) do not contain `qwen3.5`, so eval still needs
the env var or a rename.

The 2026-09-08/09 numbers were collected *before* #3780, with a patched chat
template (non-leading system rendered as ChatML system turns). That path stays
on the page so the table is reproducible; the client-side fold is a different
transform. This PR remains docs-only.

## Headline result

Base Qwen3.5-0.8B vs the markdown-trained LoRA, 116 tasks (`basic` + `all-practical`):

| format | base, free | base, grammar | SFT, free | SFT, grammar |
| --- | --- | --- | --- | --- |
| `markdown` | 12/116 (10%) | 13/116 (11%) | **22/116 (19%)** | 19/116 (16%) |
| `xml` | 9/116 (8%) | 12/116 (10%) | 5/116 (4%) | 4/116 (3%) |
| `tool` (native) | 17/116 (15%) | n/a | **26/116 (22%)** | n/a |

One epoch of SFT on your own sessions in the `markdown` format nearly doubles the
markdown pass rate, and the gain **transfers to native tool calling**. It *hurts*
the untrained format (xml went down), because in `xml` mode gptme does not parse
fenced blocks at all — format training is format-specific. Grammar-constrained
decoding helps the untrained model slightly and is not a substitute for training.

## Verification

- `python -m sphinx -b dummy docs docs/_build/dummy` — builds clean; the only
  warnings are pre-existing environment ones (`make bench-startup`, and the
  leaderboard `command-output` needing `multiprocessing_logging`), none from this page
- `docs/index.rst` toctree entry unchanged, page title unchanged, so
  https://gptme.org/docs/finetuning.html keeps working
- `git diff --check`, `prek run --files docs/finetuning.md`
- new external links (`docs.axolotl.ai`, `docs.skypilot.co`) return 200

## Background

Full experiment write-up, including the tooling verdicts, cost table and execution
log: `knowledge/strategic/2026-09-08-own-model-training-review.md` in Bob's agent
workspace (`ErikBjare/bob` — private repo, so the page describes the design rather
than linking it). Training artifacts are public at
https://s3.bob.gptme.org/training/runs/q08b-markdown-v5/

https://claude.ai/code/session_01DCzYXqioJt22SQpzoozQfQ

### CODE REVIEW

File: docs/finetuning.md [hunk 1]
```diff
@@ -1,71 +1,540 @@
...
+sessions; adding the `@tool(id): {json}` branch took the same corpus to 341.
+The "agents don't retry" conclusion drawn from the first number was a parser
+artifact. Match both:
+
+```python
+import re
+
+# fence form (markdown tool format):  ```shell\n<code>\n```
+# native form:                        @shell(abc123): {"command": "ls"}
+AT_TOOL_RE = re.compile(r"^@(\w+)\((\w+)\):\s*(\{.*\})", re.MULTILINE)
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply docs/finetuning.md]: Fixed in e80d60a34. The snippet now matches `gptme.tools.base.toolcall_re`: dotted MCP names (`server.tool`) and hyphen/colon call IDs, with `DOTALL`. `\w+` was the wrong parser.

File: docs/finetuning.md [hunk 2]
```diff
@@ -1,71 +1,540 @@
...
+`--refresh` SkyPilot answers from a local cache and will happily report a
+cluster down while the provider bills it), and back all of it with a **prepaid,
+hard-capped account**. That last one is the only cap that survives a wedged
+daemon or a bug in your own launcher.
+
+## Reproduce
+
+Five commands, in order.
+
+```bash
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply docs/finetuning.md]: Agreed — those files are not in this repo, and this PR is docs-only, so they are not being shipped. e80d60a34 makes Reproduce honest: reimplement the exporter from *Data*, and copy the checkpoint `chat_template` plus the six-line patch already on the page. The old "five commands from a clone" framing is gone.

File: docs/finetuning.md [hunk 3]
```diff
@@ -1,71 +1,540 @@
...
 ```
 
-Pick a model whose chat template matches the format you want in the resulting
-training data. The script uses `transformers.pipeline(...)` and
-`tokenizer.apply_chat_template(...)` to turn stored chats into prompt text.
-Run it in an environment where `torch` and `transformers` are already
-installed.
+- `basic` (18 tests) + `all-practical` (98) = **116 tests**, which is the
+  cheapest suite pair that still discriminates. See {doc}`evals` for the full
+  suite list.
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply docs/finetuning.md]: The 116 is measured on this repo: `basic` is 18, `all-practical` (`name.startswith("practical")`) is 98. Greptile's 100 practical / 118 total does not match `gptme.eval.suites` as of e80d60a34. The page now pins the count to 2026-09-09 on this repo and tells the reader to recount if a later revision has to match the denominator.

File: docs/finetuning.md [hunk 4]
```diff
@@ -1,71 +1,540 @@
...
+```
+
+That is also the shape the SFT data was rendered in, so training and serving
+agree. A working copy is at
+`scripts/research/training/eval/templates/qwen3_5_midsystem.jinja` in Bob's
+workspace.
+
+The proper fix belongs in gptme: fold non-leading `system` messages into the
+adjacent user turn for providers whose templates reject them, the way
+`_prep_o1` in `gptme/llm/llm_openai.py` already does for the o-series. That is
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply docs/finetuning.md]: Fixed in e80d60a34. `_prep_o1` converts each eligible system message into a *separate user* message wrapped in `<system>` tags; `_merge_consecutive` is composed with it only on the DeepSeek path. Qwen3.5 still wants a leading system message — it just cannot see another one later. The page no longer cites `_prep_o1` as the fold-into-adjacent-user model.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
## 🤖 AI code review

This PR updates the finetuning documentation to remove stale TODOs, specify the chat_template JSONL format with JSON-string arguments, describe the three renderers, and clarify eval prerequisites and execution details.

### Safe to merge — no P0/P1 findings

Confidence 5/5

✅ **No findings.** The diff looks correct to me on this pass.

<details><summary><b>Files changed (1)</b> — the diff as I read it</summary>

- `docs/finetuning.md` — _(no description)_

</details>

<details><summary>Previous review passes</summary>

| commit | score | findings | engine | when |
| --- | --- | --- | --- | --- |
| `a0c8629a6e9c` | 5/5 | 0 | llm | 2026-09-09 08:41 UTC |
| `5facb12d339d` | 5/5 | 0 | llm | 2026-09-09 08:56 UTC |
| `96b6c5d43baa` | 5/5 | 0 | llm | 2026-09-09 09:40 UTC |

</details>

<sub>Reviewed `96b6c5d43baa` · `openrouter/deepseek/deepseek-v4-flash-0731` · `llm` engine · 515s · [about this reviewer](https://github.com/gptme/gptme-contrib/blob/master/docs/ai-reviewer.md)</sub>

<details><summary><sub>Maintainer commands</sub></summary>

<sub>`@TimeToBuildBob review` (own line) — fresh review · `@TimeToBuildBob fix` — a worker acts on the findings. Once per comment; 👀 = received.</sub>

</details>

File: docs/finetuning.md [hunk 5]
```diff
@@ -1,71 +1,564 @@
...
+import re
+
+# fence form (markdown tool format):  ```shell\n<code>\n```
+# native form:                        @shell(abc123): {"command": "ls"}
+# Canonical parser: gptme.tools.base.toolcall_re
+# MCP tools are named server.tool (dots). Provider call IDs include
+# hyphens/colons (call_abc, toolu_..., call-123). `\w+` drops those.
+AT_TOOL_RE = re.compile(
+    r"^@([\w.]+)\(([\w\-:.]+)\):\s*({.*)", re.MULTILINE | re.DOTALL
+)
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply docs/finetuning.md]: Fixed in 5facb12d3. The snippet now uses `toolcall_re` + `find_json_end` and resumes after each JSON object, matching `gptme.tools.base` — the regex alone is not a parser.

File: docs/finetuning.md [hunk 6]
```diff
@@ -1,71 +1,564 @@
...
+axolotl preprocess qwen3.5-0.8b-lora.yaml
+axolotl train qwen3.5-0.8b-lora.yaml
+
+# 4. Merge the adapter into the base (vLLM serves full checkpoints, and step 3
+#    produces only an adapter), then serve. Point --chat-template at the
+#    patched copy from the section above — stock Qwen3.5 400s every gptme request.
+python3 -c "
+from peft import AutoPeftModelForCausalLM
+from transformers import AutoTokenizer
+AutoPeftModelForCausalLM.from_pretrained('/out').merge_and_unload().save_pretrained('/merged')
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply docs/finetuning.md:572]: Fixed in 5facb12d3. The yaml now sets `output_dir: /out`, matching the merge step's `from_pretrained('/out')`.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Self-merge blocked at consensus 3/5 on `5facb12d` (private exporter path, named unmerged grammar branch, no public artifact). Addressed in `96b6c5d`; remaining Greptile threads resolved. Not merging this cycle — CI is still running on the new head.

**Comment by TimeToBuildBob** [comment]
Self-merge not eligible on `96b6c5d`. Consensus gate scored 2/5 because it treated 2026-09-08/09 as a future date and Qwen3.5 / SmolLM3-3B as unverified. That is a clockless false-positive: today is 2026-09-09; both Hugging Face model cards return 200; `adapter_config.json` and `README.md` at the public run URL return 200.

CI green, Greptile 5/5, AI review 5/5 at current head, review threads resolved. Waiting for human merge.

**Comment by ErikBjare** [comment]
@TimeToBuildBob The gptme-relevant bug surfaced was fixed in https://github.com/gptme/gptme/pull/3780

**Comment by TimeToBuildBob** [comment]
Updated in `d07505632`. The page now records gptme/gptme#3780 as the client-side fix (`_fold_mid_system`; activates when the model id contains `qwen3.5`/`qwen3_5`, or `GPTME_FOLD_SYSTEM_MESSAGES=1`).

This recipe's served names (`tool-format-experiment`, `sft`) do not contain `qwen3.5`, so name inference will not fire — eval still needs the env var or a rename. The 2026-09-08/09 numbers used the patched chat template (a different transform than the fold); that path stays documented so the table is reproducible.

gptme/gptme#3779 was already closed after #3780 merged. Not merging this cycle — new head, CI running.

**Comment by TimeToBuildBob** [comment]
@greptileai review
